pad-tls/pad-tls.c

```diff
@@ -0,0 +1,112 @@
+/* pad-tls: tool to pad TLS segment of ELF files so that a given compartment's
+TLS variables will only possibly share a page with another compartment's
+padding, not its TLS variables */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <libgen.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+
+#include <elf.h>
+
+enum patch_result {
+  PATCH_SUCCESS,
+  BAD_MAGIC,
+  WRONG_ARCH,
+  NO_TLS,
+};
+
+static const char *patch_errors[] = {
+    "invalid ELF magic number",
+    "ELF architecture is not x86-64",
+    "no TLS segment in ELF program headers",
+};
+
+enum patch_result patch_tls(void *elf, int64_t delta, uint64_t *size_out) {
+  Elf64_Ehdr *ehdr = elf;
+
+  if (memcmp(ehdr->e_ident,
+             "\x7f"
+             "ELF",
+             4)) {
+    return BAD_MAGIC;
+  }
+
+  if (ehdr->e_machine != EM_X86_64) {
```

Comment:
Does this matter?

Reply:
The whole project is x86_64-only, so using this tool on an ELF with a different architecture means something is wrong.